sagemathinc/cocalc
# re-enabled email verification and extend server settings configuration

# Description

so far, this PR does

* re-enable email verification (actually, it was never really disabled, just hidden)
* present a feedback page about what's going on, especially if there are errors
* add a setting to enable/disable email related aspects in the UI (+ make progress towards configuring email in general)

what's left to do is to go through the various issues around this and figure out what they are about and fix them

[this comment](https://github.com/sagemathinc/cocalc/issues/4318#issue-549253885) in #4318 claims these are tons of issues to take care of:

* [x] https://github.com/sagemathinc/cocalc/issues/2507 → already addressed or out of scope, closing
* [x] https://github.com/sagemathinc/cocalc/issues/3148 → I'm not able to reproduce this at all. 
  * [ ] double check this when deployed in "test"
* [x] https://github.com/sagemathinc/cocalc/issues/2943 → this is about a broken jupyter notebook, not in scope of this PR
* [ ] https://github.com/sagemathinc/cocalc/issues/2707 → dup of #2505 ??? In some sense this is addressed, because the verification section in the UI is gone if email sending is disabled. However, what remains is 
     * [ ] check if the reported error happens, it doesn't hurt to show it. 
     * [ ] don't send any (welcome or other) email if the email backend is disabled
* [x] https://github.com/sagemathinc/cocalc/issues/2506 → there is a small page saying if it is successful or if there was an error
* [x] https://github.com/sagemathinc/cocalc/issues/2505 → similar to #2707 … the main take away is that there is no proper email config. a proper solution would be to make it a setting, fully configurable how the hubs send emails. that's a good thing to have, for all contexts.
* [x] https://github.com/sagemathinc/cocalc/issues/2470  → mixes up several things, probably a too large meta ticket
     * creating an annoying banner, IMHO out of scope for re-enabling this
     * without an email address, there is no verification. that's already the case.
     * if you change your email address to one that's not verified, it says so. you have to verify it. changing the address back to a verified one doesn't require re-verification.
     * analyzing bouncing/blocked emails and showing this in the UI sounds interesting, but that's way more than just enabling them again. it would probably require a new service interacting with the api and ingesting the entire stream of email activities. sounds like a ticket for its own as well.
* [ ] https://github.com/sagemathinc/cocalc/issues/886  → this is about invite codes. I don't know what this has to do with verified emails. They are mentioned, but there is no reason why.

# Testing Steps
* main flow:
   1. existing user: click the verification button in the account settings → email should arrive → click the link → see it did work → redirected to cocalc. done.
   2. new user: same as above, except there is a link in the welcome email
* problematic flow: create a new account, but do not click on the verification link. instead, copy/paste it and
  1. modify the email in the url → no such account error
  2. modify the token → shouldn't work, and email is still not verified
* trivial case: already verified, but you click on the email a second time. it tells you that you're already verified.

# Relevant Issues

### [Checklist](https://github.com/sagemathinc/cocalc/wiki/PR-Checklist):

Front end:
